### DISCUSSION TITLE
[WASM] Grpc AuthenticationException

### DISCUSSION THREAD

**Discussion by limeniye** [unoplatform/uno#11407] [Q&A] [answered]
I'm tring to send gRPC request from Wasm application.

But the next simple example is doesn't work for Wasm:
```C#
app.UseGrpcWeb();
app.MapGrpcService<GreeterService>().EnableGrpcWeb();
```

```C#
var httpClientHandler = new HttpClientHandler();
var channel = GrpcChannel.ForAddress("https://localhost:7162", new GrpcChannelOptions
{
    HttpClient = new HttpClient(new GrpcWebHandler(GrpcWebMode.GrpcWebText, httpClientHandler)),
    Credentials = ChannelCredentials.SecureSsl
});

var client = new Greeter.GreeterClient(channel);

client.SayHello(new ());
```

![зображення](https://user-images.githubusercontent.com/69314237/220391153-fb544e4f-a042-4592-87f3-eb42ca6133ed.png)

```
Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
```

Test repository: [https://github.com/limeniye/UnoPlatformGrpcAndWasm](https://github.com/limeniye/UnoPlatformGrpcAndWasm/blob/master/UnoPlatformGrpcAndWasm/UnoPlatformGrpcAndWasm/MainPage.xaml.cs#L16-L25)

- [x] Windows
- [x] Wpf Skia
- [ ] Wasm

**Comment by jeromelaban** [comment, marked as answer]
Are you able to show the actual exceptions that are happening? Their description may help troubleshooting the issue.

**Reply by limeniye** [reply]
@jeromelaban
When I'm executing next line:
```C#
client.SayHello(new ());
```
I'm expecting that my debug break point will stoped in server part.
But after this line nothing changes in backend side.

 I tried to rise some errors
(https://platform.uno/docs/articles/logging.html)

but thats all issues that I can see when I'm using Wasm

**Reply by jeromelaban** [reply]
These exceptions `Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll` are raised on the server side. Can you set the debugger to break on all exceptions and see details on those exceptions?

**Reply by limeniye** [reply]
@jeromelaban, yes, you was right. My exceptions was disabled. Thanks.

![зображення](https://user-images.githubusercontent.com/69314237/220705162-3bff35b5-8c62-4388-a0aa-1be251466f75.png)

StackTrace:
```
at System.Net.Security.SslStream.<ForceAuthenticationAsync>d__146`1.MoveNext()
```
Inner exception:
```
An unknown error occurred while processing the certificate.
```

**Reply by jeromelaban** [reply]
This looks like some self-signed certificates issue... You may want to validate that you're using the right end points, or try disabling https and see if it helps.

**Reply by limeniye** [reply]
@jeromelaban to be honest I already stuck 3 days in total and have used a lot different realizations.
But it still doesn't works for WASM.
I don't have the same problems for WinUI project, or Wpf Skia.
I've tried to use connect to HTTP but it doesn't works for me.
Maybe I'm doing something wrong.

**Reply by jeromelaban** [reply]
Do you see anything of interest in the browser debugger output?

**Reply by limeniye** [reply]
@jeromelaban, Thank you for your answers! ✅
You showed me how to rise an exceptions and I found the problem:

https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSDidNotSucceed

Resolve:
1. I changed Mozilla to Chrome

2. I added policy
```C#
const string corsPolicy = "_corsPolicy";

builder.Services.AddCors(options =>
{
    options.AddPolicy(name: corsPolicy,
        policy =>
        {
            policy.WithOrigins("https://localhost:5001",
                                "http://localhost:5000")
                .AllowAnyHeader()
                .AllowAnyMethod();
        });
});
```
```C#
app.UseCors(corsPolicy);
```
